packages/protocol/contracts/common/B12.sol

```diff
@@ -0,0 +1,898 @@
+
+  // G1Point represents a point on BLS12-377 over Fp with coordinates (X,Y);
+  struct G1Point {
+    Fp X;
+    Fp Y;
+  }
+
+  // G2Point represents a point on BLS12-377 over Fp2 with coordinates (X,Y);
+  struct G2Point {
+    Fp2 X;
+    Fp2 Y;
+  }
+
+  struct G1MultiExpArg {
+    G1Point point;
+    uint256 scalar;
+  }
+
+  struct G2MultiExpArg {
+    G2Point point;
+    uint256 scalar;
+  }
+
+  struct PairingArg {
+    G1Point g1;
+    G2Point g2;
+  }
+
+  function FpEq(Fp memory a, Fp memory b) internal pure returns (bool) {
+    return (a.a == b.a && a.b == b.b);
+  }
+
+  function fpGt(Fp memory a, Fp memory b) internal pure returns (bool) {
+    return (a.a > b.a || (a.a == b.a && a.b > b.b));
+  }
+
+  function Fp2Eq(Fp2 memory a, Fp2 memory b) internal pure returns (bool) {
+    return FpEq(a.a, b.a) && FpEq(a.b, b.b);
+  }
+
+  function fp2Gt(Fp2 memory a, Fp2 memory b) internal pure returns (bool) {
+    if (FpEq(a.b, b.b)) return fpGt(a.a, b.a);
+    else return fpGt(a.b, b.b);
+  }
+
+  function fpAdd2(uint256 a, uint256 b) internal pure returns (Fp memory) {
+    return fpAdd(Fp(0, a), Fp(0, b));
+  }
+
+  function fpAdd3(uint256 a, uint256 b, uint256 c) internal pure returns (Fp memory) {
+    return fpAdd(Fp(0, a), fpAdd(Fp(0, b), Fp(0, c)));
+  }
+
+  function fpAdd4(uint256 a, uint256 b, uint256 c, uint256 d) internal pure returns (Fp memory) {
+    return fpAdd(Fp(0, a), fpAdd(Fp(0, b), fpAdd(Fp(0, c), Fp(0, d))));
+  }
+
+  function fpAdd(Fp memory a, Fp memory b) internal pure returns (Fp memory) {
+    uint256 bb = a.b + b.b;
+    uint256 aa = a.a + b.a + (bb >= a.b && bb >= b.b ? 0 : 1);
+    return Fp(aa, bb);
+  }
+
+  function fpSub(Fp memory a, Fp memory b) internal pure returns (Fp memory) {
+    Fp memory p = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+    );
+    Fp memory x = fpAdd(a, p);
+    uint256 bb = x.b - b.b;
+    uint256 aa = x.a - b.a - (bb <= x.b ? 0 : 1);
+    return Fp(aa, bb);
+  }
+
+  function fpModExp(Fp memory base, uint256 exponent, Fp memory modulus)
+    internal
+    view
+    returns (Fp memory)
+  {
+    uint256 base1 = base.a;
+    uint256 base2 = base.b;
+    uint256 modulus1 = modulus.a;
+    uint256 modulus2 = modulus.b;
+    bytes memory arg = new bytes(3 * 32 + 32 + 64 + 64);
+    bytes memory ret = new bytes(64);
+    uint256 result1;
+    uint256 result2;
+    assembly {
+      // length of base, exponent, modulus
+      mstore(add(arg, 0x20), 0x40)
+      mstore(add(arg, 0x40), 0x20)
+      mstore(add(arg, 0x60), 0x40)
+
+      // assign base, exponent, modulus
+      mstore(add(arg, 0x80), base1)
+      mstore(add(arg, 0xa0), base2)
+      mstore(add(arg, 0xc0), exponent)
+      mstore(add(arg, 0xe0), modulus1)
+      mstore(add(arg, 0x100), modulus2)
+
+      // call the precompiled contract BigModExp (0x05)
+      let success := staticcall(gas, 0x05, add(arg, 0x20), 0x100, add(ret, 0x20), 0x40)
```

Comment:
Would be nice to not use a magic number here for the precompile address.

Ideal scenario is we have a library for using the new Byzantium precompiles (does an external library exist?) that are not built into Solidity.

Can be out of scope for this PR, but would create an issue to deal with this later.

Reply:
I did not find an external library. We could pull out all precompile addresses (both Eth and Celo) into a precompile address lib and use that in UsingPrecompiles.sol, B12.sol and CIP20Lib.sol. But then these files would not be self-contained anymore, which might be nice for users including these files in their project (since we don't provide them as library packages). I'm not sure that would be an advantage overall.

I can certainly pull out this magic number to a variable, but since the this is the only Eth precompile we use here, that also would not be clearly better than this comment.

I don't have a strong opinion on any of the approaches.

Reply:
Since this file seems to be almost copy if https://github.com/prestwich/b12-sol/blob/main/contracts/B12.sol I would personally vote for leaving it close to original source as possible.